server/src/util/my-repos.ts

```diff
@@ -0,0 +1,150 @@
+
+const MAX_STATS_RETRIES = 5;
+
+// Variables that impact the scoring algorithm.
+const SCORE_THRESHOLD = 20;
+const CONTRIBUTION_SIZE = 10;
+const CONTRIBUTION_WINDOW = 1000 * 60 * 60 * 24 * 90;  // 90 days.
+
+/**
+ * Returns a list of repos that the given user contributes to. It is ordered
+ * based on contribution recency and consistency.
+ */
+export async function getMyRepos(
+    github: GitHub, login: string, userToken: string): Promise<string[]> {
+  const repos = new Map<string, number>();
+  const results = github.cursorQuery<MyReposQuery, MyReposQueryVariables>(
+      reposQuery, {login}, (q) => q.user && q.user.contributedRepositories);
+  const promises = [];
+  for await (const data of results) {
+    if (data.user) {
+      for (const repo of data.user.contributedRepositories.nodes || []) {
+        if (repo) {
+          const promise = getContributionWeight(
+              github, repo.owner.login, repo.name, login, userToken);
+          promises.push(promise.then((score) => {
+            repos.set(repo.owner.login + '/' + repo.name, score);
+          }));
+        }
+      }
+    }
+  }
+  await Promise.all(promises);
+  const comparator = (a: string, b: string) => {
+    return (repos.get(a) || 0) < (repos.get(b) || 0) ? 1 : -1;
+  };
+  let result = Array.from(repos.keys());
+  // Filter out repos with small levels of contributions.
+  result = result.filter((key) => (repos.get(key) || 0) > SCORE_THRESHOLD);
+  return result.sort(comparator);
+}
+
+/**
+ * Given a repo and a user, returns a weighted contribution score for the repo
+ * (0-100).
+ */
+function getContributionWeight(
+    github: GitHub, org: string, repo: string, user: string, userToken: string):
+    Promise<number> {
+  return new Promise(async (resolve, reject) => {
+    try {
+      // GitHub doesn't provide a v4 API equivalent for stats, so v3 API must be
+      // used.
+      const queryPath = `repos/${org}/${repo}/stats/contributors`;
+
+      let response = await github.get(queryPath, userToken, false);
+      let retries = 0;
+      // GitHub's API may serve a cached response and begin an asynchronous
+      // job to calculate required data.
+      while (response.statusCode !== 200 && retries++ < MAX_STATS_RETRIES) {
+        await sleep(1000 * retries);
+        response = await github.get(queryPath, userToken, false);
+      }
+
+      if (response.statusCode === 200) {
+        const stats = JSON.parse(response.body) as GitHubStatsResponse;
+        // Find contribution stats for the specified user.
+        const stat = stats.filter((x) => x.author.login === user);
+        if (stat.length === 0) {
+          resolve(0);
+          return;
+        }
+
+        let contributions = 0;
+        for (const week of stat[0].weeks) {
+          // Convert UNIX timestamp in seconds to Date object.
+          const date = new Date(Number(week.w) * 1000);
+          const timeSince = Math.max(
```

Comment:
Comment (or rename the var) what's the unit for `timeSince`?

Reply:
Done.